servicetalk-concurrent-api/src/main/java/io/servicetalk/concurrent/api/RepeatWhenSingle.java

```diff
@@ -0,0 +1,192 @@
+            }
+        }
+
+        private static long sanitize(final long n) {
+            // The value must be negative because 0 is used to determine if there is no demand and a subscribe should
+            // be done. It can't overlap with our token TERMINAL or CANCEL values either.
+            return n == 0 ? -1 : max(n, MIN_INVALID_N);
+        }
+
+        @Override
+        public void cancel() {
+            for (;;) {
+                final long prev = outstandingDemand;
+                if (prev < 0 || outstandingDemandUpdater.compareAndSet(this, prev, CANCELLED)) {
+                    // prev < 0 means either we are already terminated/cancelled, or there is invalid demand pending.
+                    break;
+                }
+            }
+            sequentialCancellable.cancel();
+        }
+
+        private final class RepeatSubscriber implements SingleSource.Subscriber<T> {
+            private final CompletableSource.Subscriber completableSubscriber = new CompletableSource.Subscriber() {
+                @Override
+                public void onSubscribe(final Cancellable cancellable) {
+                    sequentialCancellable.nextCancellable(cancellable);
+                }
+
+                @Override
+                public void onComplete() {
+                    for (;;) {
+                        final long prev = outstandingDemand;
+                        assert prev != TERMINATED && prev != 0;
+                        if (prev == CANCELLED) {
+                            break;
+                        } else if (prev < 0) {
+                            // This thread owns the subscriber, no concurrency expected, no atomic necessary.
+                            onErrorInternal(newExceptionForInvalidRequestN(prev));
+                            break;
+                        } else if (outstandingDemandUpdater.compareAndSet(RepeatSubscription.this, prev, prev - 1)) {
+                            if (prev > 1) {
+                                outer.original.delegateSubscribe(RepeatSubscriber.this, contextMap, contextProvider);
+                            }
+                            break;
+                        }
+                    }
+                }
+
+                @Override
+                public void onError(final Throwable t) {
+                    outstandingDemand = TERMINATED;
+                    subscriber.onComplete(); // repeat means an error just terminates normally.
+                }
+            };
+
+            @Override
+            public void onSubscribe(final Cancellable cancellable) {
+                sequentialCancellable.nextCancellable(cancellable);
+            }
+
+            @Override
+            public void onSuccess(@Nullable final T result) {
+                final Completable completable;
+                try {
+                    subscriber.onNext(result);
+                    completable = requireNonNull(outer.repeater.apply(++redoCount, result));
```

Comment:
nit: `redoCount` looks like a leftover from the publisher impl of the same operator, consider `repeatCount`